src/platforms/esp32/components/avm_builtins/i2c_resource.c

```diff
@@ -0,0 +1,765 @@
+/*
+ * This file is part of AtomVM.
+ *
+ * Copyright 2019 Davide Bettio <davide@uninstall.it>
+ * Copyright 2023 Fred Dushin <fred@dushin.net>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later
+ */
+
+#include <sdkconfig.h>
+
+#include <string.h>
+
+// Note: The i2c.h legacy driver will be deprecated in 5.2
```

Comment:
Since this is targeting 0.7.0, and that won't likely be released in the next few months, you might save yourself the rewrite, in just a couple more months (4.4 is deprecated by Espressif at the end of July) and write this driver so it is only included in builds with ESP-IDF 5.x and use the new APIs right now.